File: src/main/kotlin/Curryfication.java
package main.kotlin;

import java.util.function.Function;

public class Curryfication {

    public static void main(String[] args) {
        Function<Integer, Function<Integer, Integer>> sum = (a) -> (b) -> a + b;

        System.out.println(sum.apply(5)
                              .apply(6));
    }
}

File: src/main/kotlin/Curryfication.kt
package main.kotlin

val sum3 = { a: Int -> { b: Int -> a + b} }

fun sum2(a: Int): (Int) -> Int = { it + a }

fun sum(a: Int) = fun(b: Int) : Int {
    return a + b
}

fun main(args: Array<String>) {
    sum2(5) (6)
    sum3(5)(6)
}
